text/0000-eager-macro-expansion.md

```diff
@@ -0,0 +1,615 @@
+it, or to re-export a hygienic symbol defined in it), the author is currently
+unable to do so.
+
+Giving proc macro authors the ability to handle these situations will allow
+proc macros to 'just work' in more contexts, and without surprising users who
+expect macro calls to interact well with more parts of the language.
+
+As a side note, allowing macro calls in built-in attributes would solve a few
+outstanding issues (see
+[rust-lang/rust#18849](https://github.com/rust-lang/rust/issues/18849) for an
+example).
+
+An older motivation to allow macro calls in attributes was to get
+`#[doc(include_str!("path/to/doc.txt"))]` working, in order to provide an
+ergonomic way to keep documentation outside of Rust source files. This was
+eventually emulated by the accepted [RFC
+1990](https://github.com/rust-lang/rfcs/pull/1990), indicating that macros in
+attributes could be used to solve problems at least important enough to go
+through the RFC process.
+
+## Interpolating macros in output
+
+Macros are currently not allowed in certain syntactic positions. Famously, they
+aren't allowed in identifier position, which makes `concat_idents!` [almost
+useless](https://github.com/rust-lang/rust/issues/29599). If macro authors have
+access to eager expansion, they could eagerly expand `concat_idents!` and
+interpolate the resulting token into their output.
+
+# Guide-level explanation
+
+We expose the following API in the `proc_macro` crate, to allow proc macro
+authors to iteratively expand all macros within a token stream:
+
+```rust
+use proc_macro::{LexError, Diagnostic, TokenStream};
+
+pub struct ExpansionBuilder { /* No public fields. */ }
+pub enum ExpansionError {
+    LexError(LexError),
+    ParseError(Diagnostic),
+    MissingDefinition(Diagnostic),
+    /* What other errors do we want to report?
+       What failures can the user reasonably handle? */
+}
+
+impl ExpansionBuilder {
+    /// Creates a new macro expansion request to iteratively expand all macro
+    /// invocations that occur in `tokens`.
+    ///
+    /// Expansion results will be interpolated within the input stream before
+    /// being returned.
+    ///
+    /// `tokens` should parse as valid Rust -- for instance, as an item or
+    /// expression.
+    pub fn from_tokens(tokens: TokenStream) -> Self;
+
+    /// Sends the expansion request to the compiler, then awaits the results of
+    /// expansion.
+    ///
+    /// The main causes for an expansion not completing right away are:
+    /// - Procedural macros performing IO or complex analysis.
+    /// - The input token stream referring to a macro that hasn't been defined
+    ///   yet.
+    pub async fn expand(self) -> Result<TokenStream, ExpansionError>;
+}
+```
+
+## Simple examples
+
+Here is an example showing how a proc macro can find out what the result of
+`concat!("hello ", "world!")` is. We assume we have access to a function
+`await_future<T>(impl Future<T>) -> T` which polls a future to completion and
```

Comment:
How would this work in practice? Can this be *any* function (e.g. `tokio::Runtime::block_on`), or is this a compiler builtin?

Since proc-macros are currently run in a different process, there needs to be some way of ensuring that `Waker::wake` can actually get called on the `Waker` that (implicitly) get passed to `Expand`

Reply:
I'll admit I'm not super knowledgeable on how futures work, but I would have thought that writing a function `Future<T> -> T` could be done with a `loop` that keeps on polling the underlying future (which is even guaranteed to make progress since we're running in a separate process). Yes, that's horribly inefficient, but that's not the point of the examples.

Would the examples be clear enough if they just used the `.await` keyword? How would one use `.await` in a (required-to-be non-asynchronous) `#[proc_macro]` function?

Reply:
Most futures have to be designed to run on some executor, or to be running in a different thread or in some context so that they can continue to run while the current future running is stopped. Usually this means passing it off to tokio or async-std, the two main executors.

I think the answer to your question depends a lot on whether you plan to make the proc macros async as well. Specifically, if proc macro functions remain synchronous, then the benefit of making this thing a Future is questionable. Usually the benefit of introducing futures is that you can run many tasks at once on a single thread, and each time one task is blocked, the others run. But if the proc_macro functions _aren't futures themselves_, then there is no benefit to having a Futures interface which calls them internally. I mean, if the compiler is going to have to start up a new thread for each proc_macro it's calling, I don't think there's a benefit to exposing an asynchronous API at all.

> How would one use .await in a #[proc_macro] function, which are currently required to be synchronous?

I think you've answered it here - it's not possible. If you want to use futures, using `.await` is going to be the most obvious way, and having the proc macros be asynchronous themselves will make this all work. If they're asynchronous, then they can just .await, and the compiler will handle all of the runtime details within itself.

If futures are to remain completely synchronous, then they can't use `.await`, and they can't use what is in my opinion the key defining feature of futures: the ability to have blocked tasks just exist as data structures, and not block entire threads.

Reply:
Alright, I think I'll just extend the RFC to also allow async `#[proc_macro]`s.

Reply:
> Since proc-macros are currently run in a different process

That's not true today, we can't even afford to run them on a separate thread atm, we may need some way to defer the RPC to lower the cost.

Reply:
Hmm, is there a feasible path forward for allowing async expansion then?

Reply:
I don't think proc-macros being run in the same process should prevent async expansion? If anything it might make it nicer.

I imagine if rustc includes a small async executor, the implementation wouldn't even need to be that expensive. The biggest part of the executor is for handling async io, and if we're really just handling a bunch of different tasks which depend on eachother, then a super simple executor could work.

I'm not at all sure how we would build the expand() future, and how we resolve expansion errors in a predictable & reliable way, but the fact of it being async shouldn't be a big complication.